### ISSUE TITLE
Media: Local previews when uploading unsupported file type

### ISSUE THREAD

**Issue by swissspidy** [GoogleForCreators/web-stories-wp#6088]
## Feature Description

When uploading videos and images, we try to display them on the canvas or on the library so the user can already interact with it while uploading.

When uploading unsupported file types, it is currently not possible to do that. There can even be errors (see #6310)

To solve this, we will do three things, in this order:

1. Display video placeholder
2. Try to generate a resource from the local file (`getResourceFromLocalFile`, using `createObjectURL`)
3. If that fails (i.e. the file type is unsupported), try poster generation using ffmpeg

**Important:** this is not just useful for video optimization, as (1) and (2) are relevant even when uploading MP4s.

###  Video Placeholder (new)

This is the fallback solution when uploading videos. The first thing we do. We display a gray rectangle with a disabled play button. A tooltip explains the user why it is grayed out.

**Example:**

<img width="309" alt="Screenshot 2021-02-12 at 12 51 09" src="https://user-images.githubusercontent.com/841956/107764731-fc14a000-6d30-11eb-985c-e3c829a50fd6.png">

**Note:** Exact implementation TBD. Whether this should be an actual gray rectangle with an icon, or an actual video tag with some styling & missing `source` needs to be tested/discovered.

### Display video/poster using blobs (existing)

This is the current solution we have **today**. Try to load the video and a poster using `createObjectURL` and `getFirstFrameOfVideo`

**Note:** Exact implementation TBD. Whether this should be an actual gray rectangle with an icon, or an actual video tag with some styling & missing `source` needs to be tested/discovered.

### Display poster using `ffmpeg` (new)

**Handled in #6399**

<details>
<summary>Details</summary>

This is a fallback solution in case the above doesn't work.

Use `ffmpeg` to generate a poster image from the video that can be showed instead of the placeholder.

We show the poster image  with a disabled play button

Docs: https://trac.ffmpeg.org/wiki/Create%20a%20thumbnail%20image%20every%20X%20seconds%20of%20the%20video

**Example:**

<img width="225" alt="Screenshot 2021-02-12 at 13 04 02" src="https://user-images.githubusercontent.com/841956/107765907-ee601a00-6d32-11eb-9657-b93bf3bec7c5.png">

</details>

### Tooltip (new)

As the screenshot above shows, we display a tooltip when the user hovers over the video while the video is being progressed.

This tooltip is shown regardless of whether we display a placeholder or a poster.

This can also be implemented in a separate ticket.

## Additional Context

As a reminder, the flow is like this:

User drops video -> `useUploadMedia` -> `useUploader` (calls `useTranscode`) -> upload to WordPress

The logic for displaying local files in the library is handled in `useUploadMedia` (`assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js`). Seems like something there needs to be adjusted to make it work for video optimization.

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

**Comment by spacedmonkey** [comment]
The current behaviour is in place to preview a local version, what we know for sure will play in most browsers ( mp4 / webp ). What if the user uploads .mov or other file format is that the browser can not display out of the box? Is https://github.com/google/web-stories-wp/issues/6053 a blocker for this ticket?

**Comment by swissspidy** [comment]
I think it is, yes.

**Comment by spacedmonkey** [comment]
From #6310

TLDR
> if the current method (createObjectURL) does not work, we try to generate a poster with ffmpeg. 

Pascal brings up an interesting point, regarding the time for poster image to generate. 

> Now, that poster generation is gonna take some time (let's say 30 seconds, just for the sake of it). During that time, there still needs to be something we can display to the user - a placeholder.

So the user story would be this. 

- User uploads unsupported video file
- transcoding kicks in. 
- Video element created with placeholder ( loading bar maybe? ).
- ffmpeg is used to generate poster image ( this takes 10 seconds ). 
- ffmpeg transcodes the video ( this takes 1 minute ). 

So at 0 seconds, placeholder, 10 seconds poster image, 60 seconds, usable video. This makes sense to me, as it gives the user something to interact with. But we need to workout timings. If it is - 0 seconds placeholder, 50 seconds poster, 60 seconds video, it is a little pointless / less useful for the user. 

We need to understand 2 things. 
- How long does it take to generate a poster image using ffmpeg? 
- Does it slow the transcoding down by generating the poster image?

At a guess, I would say no. Both would require ffmpeg be to loaded, so at the very lease, we need for that to load and init. But after that, generating a image SHOULD be A LOT quicker than transcoding a whole video.

**Comment by swissspidy** [comment]
> Does it slow the transcoding down by generating the poster image?

As I mentioned in https://github.com/google/web-stories-wp/issues/6310#issuecomment-777666461, this poster generation using ffmpeg can happen **in parallel** to the actual transcoding.

----

NB: Since we first and foremost need a placeholder, and the ffmpeg part is a nice progressive enhancement to that, we could split this up into two tickets

**Comment by spacedmonkey** [comment]
> As I mentioned in #6310 (comment), this poster generation using ffmpeg can happen in parallel to the actual transcoding.

Yeah, I know, they are both promise based and can return when they return. But poster generate will (most likely) finish first. 

> NB: Since we first and foremost need a placeholder, and the ffmpeg part is a nice progressive enhancement to that, we could split this up into two tickets

Good idea.

**Comment by swissspidy** [comment]
Created #6399 for the ffmpeg enhancement so we can focus on the placeholder part.

I hope we can simply use a `<video>` element for that with a missing source and add some styling. If not.

**Comment by csossi** [comment]
"MOV video (H.264), dropped on editor

Drop MOV video onto editor
See an instant preview of the video on the canvas and in the media library
Verify that you can play the video on the canvas while it's being uploaded
See a progress bar for the video in the media library while it's being uploaded
Verify that you get a warning trying to close the window while upload is in progress"

Grey placeholder is appearing - unable to play video while it is uploading
![image.png](https://images.zenhubusercontent.com/5ea085e4704e72c0a99c6997/91c89b23-8236-4550-a6d3-03db2bd16fcd)

**Comment by swissspidy** [comment]
> Grey placeholder is appearing - unable to play video while it is uploading

@csossi I'll double check, but perhaps my testing instruction was a bit off there. If everything else was fine this should be acceptable behavior.

**Comment by swissspidy** [comment]
@csossi confirmed that my testing instruction was out of date. gray placeholder on canvas is expected

**Comment by csossi** [comment]
Thanks, @swissspidy 

Verified in QA
